src/v/archival/async_data_uploader.h

```diff
@@ -0,0 +1,178 @@
+struct inclusive_offset_range {
+    model::offset base;
+    model::offset last;
+
+    inclusive_offset_range() = default;
+
+    inclusive_offset_range(model::offset base, model::offset last)
+      : base(base)
+      , last(last) {
+        vassert(base <= last, "Invalid offset range {} - {}", base, last);
+    }
+
+    int64_t size() const noexcept { return last() - base() + 1; }
+
+    bool contains(const model::record_batch_header& hdr) const noexcept {
+        return base <= hdr.base_offset && hdr.last_offset() <= last;
+    }
+
+    auto begin() const noexcept {
+        auto cit = boost::make_counting_iterator<int64_t>(base());
+        return boost::make_transform_iterator(cit, &detail::cast_to_offset);
+    }
+
+    auto end() const noexcept {
+        auto cit = boost::make_counting_iterator<int64_t>(last() + 1);
+        return boost::make_transform_iterator(cit, &detail::cast_to_offset);
+    }
+};
+
+inline std::ostream& operator<<(std::ostream& o, inclusive_offset_range r) {
+    fmt::print(o, "[{}-{}]", r.base, r.last);
+    return o;
+}
+
+/// Result of the upload size calculation.
+/// Contains size of the region in bytes and locks.
+struct cloud_upload_parameters {
+    /// Size of the upload in bytes
+    size_t size_bytes;
+    /// True if at least one segment is compacted
+    bool is_compacted;
+    /// Offset range of the segment
+    inclusive_offset_range offsets;
+    /// First offset inside the uploaded segment (might be different from
+    /// metadata offset if the segment is compacted)
+    model::offset first_real_offset;
+    /// Timestamp from the first data batch
+    model::timestamp first_data_timestamp;
+    /// Last offset inside the uploaded segment (might be different from
+    /// metadata offset if the segment is compacted)
+    model::offset last_real_offset;
+    /// Timestamp from the last data batch
+    model::timestamp last_data_timestamp;
+    /// Segment level read locks
+    std::vector<ss::rwlock::holder> segment_locks;
+};
+
+/// Individual segment upload
+class segment_upload {
+public:
+    explicit segment_upload(
+      model::ntp ntp,
+      inclusive_offset_range offsets,
+      size_t read_buffer_size,
+      ss::scheduling_group sg);
+
+    /// Initialize segment upload
+    ss::future<result<void>> initialize(
+      ss::lw_shared_ptr<cluster::partition> part,
+      model::timeout_clock::duration timeout);
+
+    /// Close upload and free resources
+    ss::future<> close();
+
+    /// Get input stream or throw
+    ss::input_stream<char> get_stream() && {
+        throw_if_not_initialized("get_stream");
+        auto tmp = std::exchange(_stream, std::nullopt);
+        return std::move(tmp.value());
+    }
+
+    /// Get precise content length for the upload
+    size_t get_size_bytes() const {
+        throw_if_not_initialized("get_size_bytes");
+        return _params->size_bytes;
+    }
+
+private:
+    void throw_if_not_initialized(std::string_view caller) const;
```

Comment:
I assume you're splitting the constructor and initialization because you need initialization to be async? Why not make the constructor private and have a static async function that constructs the object and initializes it properly? This way you wouldn't have to add guards everywhere and will be impossible to have a "non-initialized" `segment_upload` object.